xmodule/library_sourced_block.py

```diff
@@ -132,28 +253,126 @@ def validate(self):
+        return values
+
+    @property
+    def non_editable_metadata_fields(self):
+        non_editable_fields = super().non_editable_metadata_fields
+        non_editable_fields.extend([
+            LibrarySourcedBlock.source_block_ids,
+            LibrarySourcedBlock.source_library_version,
+        ])
+        return non_editable_fields
+
+    @property
+    def source_library_key(self):
         """
-        Save changes to this block, applying edits made in Studio.
+        Convenience method to get the library ID as a LibraryLocator and not just a string
         """
-        response = super().submit_studio_edits(data, suffix)
-        # Replace our current children with the latest ones from the libraries.
-        lib_tools = self.runtime.service(self, 'library_tools')
         try:
-            lib_tools.import_from_blockstore(self, self.source_block_ids)
-        except Exception as err:  # pylint: disable=broad-except
-            log.exception(err)
-            return Response(_("Importing Library Block failed - are the IDs valid and readable?"), status=400)
-        return response
+            return LibraryLocator.from_string(self.source_library_id)
+        except InvalidKeyError:
+            return LibraryLocatorV2.from_string(self.source_library_id)
+
+    @lazy
+    def tools(self):
+        """
+        Grab the library tools service or raise an error.
+        """
+        return self.runtime.service(self, 'library_tools')
+
+    def editor_saved(self, user, old_metadata, old_content):  # lint-amnesty, pylint: disable=unused-argument
+        """
+        If source_library_id is empty, clear source_library_version and children.
+        """
+        if not self.source_library_id:
+            self.children = []  # lint-amnesty, pylint: disable=attribute-defined-outside-init
+            self.source_library_version = ""
+        else:
+            self.source_library_version = str(self.tools.get_library_version(self.source_library_id))
+
+    def post_editor_saved(self):  # lint-amnesty, pylint: disable=unused-argument
+        """
+        If source_library_id has been edited, refresh_children automatically.
+        """
+        try:
+            if self.source_library_id:
+                self.refresh_children()
+        except ValueError:
+            pass  # The validation area will display an error message, no need to do anything now.
+
+    @XBlock.handler
+    def refresh_children(self, request=None, suffix=None):  # lint-amnesty, pylint: disable=unused-argument
+        """
+        Refresh children:
+        This method is to be used when any of the libraries that this block
+        references have been updated. It will re-fetch all matching blocks from
+        the libraries, and copy them as children of this block. The children
+        will be given new block_ids, but the definition ID used should be the
+        exact same definition ID used in the library.
```

Comment:
> but the definition ID used should be the exact same definition ID used in the library.

That's not possible for blockstore blocks imported into modulestore, as they use different formats for the definition IDs. I think this is leftover from an earlier version.

Reply:
https://github.com/openedx/edx-platform/pull/33263#pullrequestreview-1627813331